### PR TITLE
feat: secrets enabled for rdws services

### PR DESCRIPTION
**Pull request by yohanthshetty**
## Overview:
In this PR, we are adding Secrets(ssm parameter store and secrets manager secrets) for rdws(app runner services).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/workload.go
Comment (yohanthshetty) [inline internal/pkg/template/workload.go]: New variable RuntimeEnvironmentSecrets added.

File: internal/pkg/template/workload.go
Comment (yohanthshetty) [inline internal/pkg/template/workload.go]: hasRuntimeEnvironmentSecrets parsing function to check if rdws service Manifest has secrets

File: internal/pkg/manifest/rd_web_svc.go
Comment (iamhopaul123) [inline internal/pkg/manifest/rd_web_svc.go]: Why not just 
```suggestion
	Secrets         map[string]Secret                    `yaml:"secrets"`
```

File: internal/pkg/aws/apprunner/service.go
Comment (iamhopaul123) [inline internal/pkg/aws/apprunner/service.go]:
```suggestion
	Value string
```

File: internal/pkg/aws/apprunner/service.go
Comment (iamhopaul123) [inline internal/pkg/aws/apprunner/service.go]: Can we add doc string for public struct?
```suggestion
type EnvironmentSecret struct {
```

File: internal/pkg/template/workload.go
Comment (iamhopaul123) [reply internal/pkg/template/workload.go]: Can we remove `Runtime` throughout the PR? Or is there any reason why we want to add it?

Also why not reuse the existing `Secret` in `WorkloadOpts` for request driven web service (rdws)?

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (iamhopaul123) [inline internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: We might only want to include this additional policy if they have secrets enabled? aka `hasSecrets`

File: internal/pkg/template/templates/workloads/services/rd-web/cf.yml
Comment (iamhopaul123) [inline internal/pkg/template/templates/workloads/services/rd-web/cf.yml]: indentation needs to be consistent with variables section below and we don't need `-` because it doesn't create extra spaces as they are in the same line.
```suggestion
              - Name: {{$name}}
                Value: {{if not $secret.RequiresSub }}{{$secret.ValueFrom}}{{else}} !Sub 'arn:${AWS::Partition}:{{$secret.Service}}:${AWS::Region}:${AWS::AccountId}:{{$secret.ValueFrom}}' {{end }}
```

**Review by iamhopaul123 — COMMENTED** [review]
Thank you for your contribution! Can we add unit tests and [local integ test](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/deploy/cloudformation/stack/rd_web_svc_integration_test.go) for this feature? Example integ test will be like [this](https://github.com/aws/copilot-cli/commit/41223e60243bd32460eb0ca32243201a34f3a8c4#diff-8a5e86d5e588a90aef0eb7fe24342476db01a0b639bb2a9cc7db23e9cc4505f1)

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: How do I include this? Withinn if statements?
{{- if hasSecrets .}}
    - PolicyName: !Join ['', [!Ref AppName, '-', !Ref EnvName, '-', !Ref WorkloadName, SecretsPolicy]]......
{{- end }}

Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: yes something like that if you use `hasSecrets` for rdws as well.

File: internal/pkg/aws/apprunner/service.go
Comment (yohanthshetty) [reply internal/pkg/aws/apprunner/service.go]: Added to new commit.

File: internal/pkg/aws/apprunner/service.go
Comment (yohanthshetty) [reply internal/pkg/aws/apprunner/service.go]: Updated in recent commit

File: internal/pkg/manifest/rd_web_svc.go
Comment (yohanthshetty) [reply internal/pkg/manifest/rd_web_svc.go]: Done

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: Done. Updated in next commit.

File: internal/pkg/template/templates/workloads/services/rd-web/cf.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/services/rd-web/cf.yml]: Done.

File: internal/pkg/template/workload.go
Comment (yohanthshetty) [reply internal/pkg/template/workload.go]: Ack. I have reused Secret

File: internal/pkg/aws/apprunner/service.go
Comment (yohanthshetty) [inline internal/pkg/aws/apprunner/service.go:38]: // EnvironmentSecret contains the name and value of a Secret from SSM Parameter Store or Secrets Manager.
Will add the above doc string in next commit. Missed in this commit.

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: It's ok we'll squash and merge the PR after it gets approved, so don't worry about spamming commits.

**Review by iamhopaul123 — APPROVED** [review]
Awesome! This is much better :shipit:

File: internal/pkg/template/templates/workloads/services/rd-web/manifest.yml
Comment (Lou1415926) [inline internal/pkg/template/templates/workloads/services/rd-web/manifest.yml:68]: Nice! Helpful for users to discover this feature 👍🏼

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (Lou1415926) [inline internal/pkg/template/templates/workloads/partials/cf/instancerole.yml:59]: We probably want to add `Condition` to these policy statements like this: https://github.com/aws/copilot-cli/blob/4588b2227fc049a8364a19fcbc9f2c0ad604e697/internal/pkg/template/templates/workloads/partials/cf/executionrole.yml#L35

This way, Copilot will be authorized to fetch the SSM/secretsmanager value only if the SSM/secretsmanager is tagged with:
```
copilot-application: <appname>
copilot-environment: <envname>
```

So that Copilot is granted the minimal permissions that it needs

File: internal/pkg/template/templates/workloads/services/rd-web/cf.yml
Comment (Lou1415926) [inline internal/pkg/template/templates/workloads/services/rd-web/cf.yml:85]: Do you think we can reuse this [secrets partial file](https://github.com/aws/copilot-cli/blob/4588b2227fc049a8364a19fcbc9f2c0ad604e697/internal/pkg/template/templates/workloads/partials/cf/) by adding `{{include "secrets" . }}` here instead?  Or am I missing something  🤔 ?

File: internal/pkg/template/templates/workloads/services/rd-web/manifest.yml
Comment (Lou1415926) [inline internal/pkg/template/templates/workloads/services/rd-web/manifest.yml]:
```suggestion
#  GITHUB_TOKEN: GITHUB_TOKEN  # The key is the name of the environment variable, the value is the name of an SSM parameter.
```

File: internal/pkg/deploy/cloudformation/stack/testdata/workloads/rdws-manifest.yml
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/stack/testdata/workloads/rdws-manifest.yml:60]: nit: Can we also do `from_cfn` and `secretsmanager` to see how it renders?
```suggestion
    secrets:                      # Pass secrets from AWS Systems Manager (SSM) Parameter Store and Secrets Manager.
      my-ssm-secret: "arn:aws:ssm:us-east-1:123456789123:parameter/jan11ssm"
      MY_FROM_CFN_SECRET:
        from_cfn: my-other-stack-exported-secret
      MY_SECRETS_MANAGER_SECRET:
        secretsmanager: mysecretsmanagername
```

File: internal/pkg/describe/rd_web_service_test.go
Comment (Lou1415926) [inline internal/pkg/describe/rd_web_service_test.go:156]: 🤫

**Review by Lou1415926 — COMMENTED** [review]
Thank you for your contribution! This looks great ❤️ I just have a few nits - after addressing them we can :shipit:

File: internal/pkg/deploy/cloudformation/stack/testdata/workloads/rdws-manifest.yml
Comment (yohanthshetty) [reply internal/pkg/deploy/cloudformation/stack/testdata/workloads/rdws-manifest.yml:60]: Addressed in next commit. Renders fine!

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml:59]: I removed the condition for tags intentionally as we don't need these tags for secrets with App Runner service. Is it absolutely necessary to have them?

File: internal/pkg/template/templates/workloads/services/rd-web/cf.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/services/rd-web/cf.yml:85]: Good call out. I have intentionally written this field definition in the cf.yml because for _RDWS_ service the CFN template should have **RuntimeEnvironmentSecrets** as the field heading instead of **Secrets**(which is the case for _ECS_ services). And also the Value field is **Value** for _RDWS_ service and **ValueFrom** for _ECS_ service.

File: internal/pkg/template/templates/workloads/services/rd-web/manifest.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/services/rd-web/manifest.yml]: Addressed in next commit.

File: internal/pkg/describe/rd_web_service_test.go
Comment (yohanthshetty) [reply internal/pkg/describe/rd_web_service_test.go:156]: It's a separate test so I used two unique secrets for easier differentiation. Does not affect functionality in any way.

File: internal/pkg/describe/service.go
Comment (Lou1415926) [inline internal/pkg/describe/service.go]: nit: perhaps calling `SliceStable` twice gives us the result that we want for now, but it might not in the future because go lib is free to change whatever algorithm they use under the hood. I think this ⬇️ should guarantee to get what's intended here
```suggestion
	sort.SliceStable(e, func(i, j int) bool { 
	  if e[i].Name == e[j].Name {
	     return  e[i].Environment < e[j].Environment
	  }
	return e[i].Name < e[j].Name 
	})
```

**Review by Lou1415926 — DISMISSED** [review]
Looks great to me! I will approve the PR once the feedbacks are addressed in the next commit 👍🏼

File: internal/pkg/aws/apprunner/apprunner_test.go
Comment (efekarakus) [inline internal/pkg/aws/apprunner/apprunner_test.go:84]: nit: can we add another value here to test the sorting? for example `parameter/zzz123` to make sure it appears after `jan11ssm`

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (efekarakus) [inline internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: I wonder if we are at risk of hitting the 128 char limit here for the policy name: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-policy.html#cfn-iam-policy-policyname

What do you think of simply:
```suggestion
        - PolicyName: 'AccessCopilotTaggedSecrets'
```

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (efekarakus) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml:59]: I can add context in the mean time and Wanxian can provide more data if needed.

It's necessary to have the tags otherwise we're giving v broad permissions to app runner services created with Copilot. Today, Copilot ensures that users opt-in to reading secrets by forcing them to add `copilot-application` and `copilot-environment` tags to their secrets (see https://aws.github.io/copilot-cli/docs/developing/secrets/). To make the process easy, the `copilot secret init` command automatically does that.

File: internal/pkg/describe/service.go
Comment (efekarakus) [inline internal/pkg/describe/service.go]: Can we rename this struct to `rdwsSecret` since it's really similar to the existing `secret` type which also has an "environment" field:
https://github.com/aws/copilot-cli/blob/877347fa4f8f1112d7f5b7cb18c6726b3aa2d8e4/internal/pkg/describe/service.go#L478-L485

File: internal/pkg/aws/apprunner/apprunner_test.go
Comment (yohanthshetty) [reply internal/pkg/aws/apprunner/apprunner_test.go:84]: I added that, and checked. Test case Passed!

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml]: Ack.

File: internal/pkg/describe/service.go
Comment (yohanthshetty) [reply internal/pkg/describe/service.go]: Updated in next commit.

File: internal/pkg/template/templates/workloads/partials/cf/instancerole.yml
Comment (yohanthshetty) [reply internal/pkg/template/templates/workloads/partials/cf/instancerole.yml:59]: As discussed, I have added the Condition to check for tagged secrets.

File: internal/pkg/describe/service.go
Comment (yohanthshetty) [reply internal/pkg/describe/service.go]: Done!

**Review by efekarakus — APPROVED** [review]

**Comment by paragbhingre** [comment]
#4506

**Pull request merged**
